### ISSUE TITLE
How to modify the activation function?

### ISSUE THREAD

**Issue by zxsitu** [ultralytics/yolov5#3013]
## ❔Question
Hello author, I have seen that there are new activation functions added to the program, but I'm not quite sure if I've modified the code correctly, and I'd like you to give me some advice.

## Additional context
I see that you have given the prepared lines of code under this question, but I am a bit confused: 
https://github.com/ultralytics/yolov5/blob/c9c95fb282322bc2928f44488f2467c5f4104f09/models/common.py#L34-L51

1. How should FRelu be added to it? What is the format?
`self.act = nn.FRelu() if act is True else (act if isinstance(act, nn.Module) else nn.Identity())` or
`self.act = FRelu() if act is True else (act if isinstance(act, nn.Module) else nn.Identity())`
2. Why some activation functions start with nn.xxx, while some directly start with the name of the activation function? Should I use the former or the latter?

**Comment by glenn-jocher** [comment]
@ilem777 see Conv() module in activations study branch for example implementations of alternative activation functions:
https://github.com/ultralytics/yolov5/blob/0824388b9e1afb5a888ce4c302acfe2ad3da8101/models/common.py#L34-L57

**Comment by ilem777** [comment]
@glenn-jocher Hello, I have used the [study_activation branch](https://github.com/ultralytics/yolov5/tree/study_activations) and unlocked the line as you suggested
https://github.com/ultralytics/yolov5/blob/0824388b9e1afb5a888ce4c302acfe2ad3da8101/models/common.py#L47
however, I encountered the following error report. Currently I want to use the Mish activation function. How can I solve this problem?
```
Traceback (most recent call last):
  File "train.py", line 544, in <module>
    train(hyp, opt, device, tb_writer)
  File "train.py", line 88, in train
    model = Model(opt.cfg or ckpt['model'].yaml, ch=3, nc=nc, anchors=hyp.get('anchors')).to(device)  # create
  File "F:\PycharmProjects\yolov5_env\yolov5-mish\models\yolo.py", line 85, in __init__
    self.model, self.save = parse_model(deepcopy(self.yaml), ch=[ch])  # model, savelist
  File "F:\PycharmProjects\yolov5_env\yolov5-mish\models\yolo.py", line 242, in parse_model
    m_ = nn.Sequential(*[m(*args) for _ in range(n)]) if n > 1 else m(*args)  # module
  File "F:\PycharmProjects\yolov5_env\yolov5-mish\models\common.py", line 181, in __init__
    self.conv = Conv(c1 * 4, c2, k, s, p, g, act)
  File "F:\PycharmProjects\yolov5_env\yolov5-mish\models\common.py", line 47, in __init__
    self.act = Mish() if act is True else (act if isinstance(act, nn.Module) else nn.Identity())
NameError: name 'Mish' is not defined
```

**Comment by ilem777** [comment]
![image](https://user-images.githubusercontent.com/61718945/116841514-0fd2f280-ac0c-11eb-8cdb-1c82a24f1ae7.jpeg)
Do i have to install another package?

**Comment by besmaguesmi** [comment]
Hi, @glenn-jocher  why did you replace the Relu activation function with the sigmoid function in the last version? I'm really curious to understand the results that provides you ( I don't have the time to try all the parameters that's why I'm asking XD)
Thanks.

**Comment by glenn-jocher** [comment]
@besmaGuesmi architecture updates are typically informed by empirical results of experiments and studies we run. You can see our Activations Study at https://wandb.ai/glenn-jocher/activations, and a discussion at https://github.com/ultralytics/yolov5/issues/2891

**Comment by besmaguesmi** [comment]
@glenn-jocher what I've seen is that the FReLU provides the best result! why didn't you choose it?

**Comment by glenn-jocher** [comment]
@besmaGuesmi FReLU may be suitable for smaller models like 5n and 5s, but it adds too many operations to larger models and causes earlier overfitting. It also requires substantially increased resources like CUDA memory, which is not compatible with our goal of good results on consumer hardware using less resources.

**Comment by besmaguesmi** [comment]
Understood, could you please tell me how can I change the activation function in the model, I really like to use the FReLU function instead of SiLU because I used the Yolov5n model.

Thanks.

**Comment by besmaguesmi** [comment]
Is it enough to change it only in the common.py and expriment.py files?

**Comment by glenn-jocher** [comment]
@besmaGuesmi activations are defined in one place for all official YOLOv5 models:
https://github.com/ultralytics/yolov5/blob/79bca2bf64da04e7e1e74a132eb54171f41638cc/models/common.py#L38

**Comment by besmaguesmi** [comment]
@glenn-jocher  I tried the FReLU function as below
```
`
class FReLU(nn.Module):
    def __init__(self, c1, k=3):  # ch_in, kernel
        super().__init()__()
        self.conv = nn.Conv2d(c1, c1, k, 1, 1, groups=c1)
        self.bn = nn.BatchNorm2d(c1)

    @staticmethod
    def forward(self, x):
        return torch.max(x, self.bn(self.conv(x)))

class Conv(nn.Module):
    # Standard convolution
    def __init__(self, c1, c2, k=1, s=1, p=None, g=1, act=True):  # ch_in, ch_out, kernel, stride, padding, groups
        super().__init__()
        self.conv = nn.Conv2d(c1, c2, k, s, autopad(k, p), groups=g, bias=False)
        self.bn = nn.BatchNorm2d(c2)
        self.act = nn.FReLU() if act is True else (act if isinstance(act, nn.Module) else nn.Identity())`
```

But there is an error in the training! are there any mistakes in the implementation above? Thanks

**Comment by glenn-jocher** [comment]
@besmaGuesmi see https://github.com/ultralytics/yolov5/issues/3013#issuecomment-830852451

**Comment by besmaguesmi** [comment]
Hi @glenn-jocher, When I tried to change the Conv class, there are some issues because of the same files, what I understood is I have to change some files content according to https://github.com/ultralytics/yolov5/tree/0824388b9e1afb5a888ce4c302acfe2ad3da8101, but is there any other method to use directly FReLU without the need to change files like general.py, utils.py, etc.

Thanks

**Comment by glenn-jocher** [comment]
@besmaGuesmi the only file you need to update is common.py, you just import and use FReLU as in https://github.com/ultralytics/yolov5/issues/3013#issuecomment-830852451

**Comment by besmaguesmi** [comment]
@glenn-jocher yes I did exactly the same work but I faced an error in the training, do I have to remove the forward_fuse function from the Conv? 
![image](https://user-images.githubusercontent.com/36562278/141019690-99f1e8a2-a122-4912-9d2f-c0e91b9815e6.png)

**Comment by glenn-jocher** [comment]
@besmaGuesmi your python indentations are incorrect. This is unrelated to YOLOv5. You may want to take a beginner's python course first to learn the basics.

**Comment by besmaguesmi** [comment]
sorry, I uploaded the wrong screenshot, I talked about this one, what I understand is that when I cloned yolov5, the version is not updated so we have to change other python files (activation.py, general.py ...) in addition, when I checked the common.py it also missed the activation function importation (FReLU, FReLU_noBN_biasFalse, FReLU_noBN_biasTrue..) the common.py, the activation.py, etc I have obtained after cloned are not the same here:https://github.com/ultralytics/yolov5/tree/0824388b9e1afb5a888ce4c302acfe2ad3da8101/models. Did you understand what I mean by my first question?
![image](https://user-images.githubusercontent.com/36562278/141019790-ac2f74ac-77c3-4b90-96c1-cd80eddca564.png)

**Comment by besmaguesmi** [comment]
Solved by removing the other activation function, thanks

**Comment by ppogg** [comment]
Sir, I will experiment with the new activation function, as well as a lighter backbone, etc. If there is progress, I will let you know～

**Comment by fanghua2021** [comment]
Hello, I used silu, h-swish, and leaky relu respectively in yolov5s for experiments. The results show: (1) map: h-swish>swish>leaky relu; (2) FPS: silu=leaky relu> h- swish>.

I have a question that h-swish does not use an exponential function, shouldn't the speed be greater than silu? In addition, I see that FReLU’s map is the best, so is his speed the best in yolov5s?

**Comment by fanghua2021** [comment]
Hello, I used silu, h-swish, and leaky relu respectively in yolov5s for experiments. The results show: (1) map: h-swish>swish>leaky relu; (2) FPS: silu=leaky relu> h- swish>.

I have a question that h-swish does not use an exponential function, shouldn't the speed be greater than silu? In addition, I see that FReLU’s map is the best, so is his speed the best in yolov5s?
 [
![image](https://user-images.githubusercontent.com/80299078/144709903-765dcd4d-c235-4dd4-bda4-b331f5c145aa.png)
](url)

**Comment by glenn-jocher** [comment]
@fanghua2021 architecture updates are typically informed by empirical results of experiments and studies we run. You can see our Activations Study with YOLOv5s on COCO for 300 epochs at https://wandb.ai/glenn-jocher/activations, and a discussion at https://github.com/ultralytics/yolov5/issues/2891

Results may vary by dataset and model.

**Comment by hellodennis4** [comment]
Hi @glenn-jocher,Dy-relu,how do you think?it might be work better.

**Comment by glenn-jocher** [comment]
@hellodennis4 you can see our Activations Study with YOLOv5s on COCO for 300 epochs at https://wandb.ai/glenn-jocher/activations, and a discussion at https://github.com/ultralytics/yolov5/issues/2891

Results may vary by dataset and model.

**Comment by XhHello** [comment]
> Hello, I used silu, h-swish, and leaky relu respectively in yolov5s for experiments. The results show: (1) map: h-swish>swish>leaky relu; (2) FPS: silu=leaky relu> h- swish>.
> 
> I have a question that h-swish does not use an exponential function, shouldn't the speed be greater than silu? In addition, I see that FReLU’s map is the best, so is his speed the best in yolov5s? [ ![image](https://user-images.githubusercontent.com/80299078/144709903-765dcd4d-c235-4dd4-bda4-b331f5c145aa.png) ](url)

Friend, how did you draw this curve

**Comment by marziyemahmoudifar** [comment]
> @ilem777 see Conv() module in activations study branch for example implementations of alternative activation functions:
> 
> https://github.com/ultralytics/yolov5/blob/0824388b9e1afb5a888ce4c302acfe2ad3da8101/models/common.py#L34-L57

How do I add a new ELU activation function to yolov5 and use it?

**Comment by glenn-jocher** [comment]
@marziyemahmoudifar you can simply replace the default nn.SiLU() activation here on models.py L44 with another one of your design. This will affect all activations in the whole YOLOv5 model:
https://github.com/ultralytics/yolov5/blob/2e57b8418db137baafd38e71f885632badae4138/models/common.py#L38-L45

**Comment by bzha5848** [comment]
Hi, I think it still doesn't work for FReLU, when I follow the comments and modify it the colab still says no module called FReLU

**Comment by bzha5848** [comment]
![image](https://user-images.githubusercontent.com/49177682/184964094-4d889130-52d5-4afb-8c2d-ef0a4745070b.png)

**Comment by glenn-jocher** [comment]
@bzha5848 you can import FReLU from utils.activations

**Comment by ZhixiongSun** [comment]
@glenn-jocher Hi I have seen the result you mentioned here,  https://wandb.ai/glenn-jocher/activations. For FRelu-noBN-BiasTrue and FRelu-noBN-BiasFalse, it semms like you early stop it. Could you please tell me why you did this or are these two type of activations（FRelu-noBN-BiasTrue and FRelu-noBN-BiasFalse） better than original FRelu

**Comment by glenn-jocher** [comment]
@ZhixiongSun I don’t remember exactly,  but typical early stopping reasons may be excess resource usage, ie CUDA memory, or slow training speed.

You should be able to reproduce the runs yourself using the commands shown in the wandb logs.

**Comment by passerbythesun** [comment]
Give a DP about `FReLU`:
I've tested `FReLU`(blue line) in our custom dataset, just as describe aboved, the result is not good. The training procedure stopped at epoch 145, since "no improvement observed in last 100 epochs". 
![image](https://github.com/ultralytics/yolov5/assets/3233540/e88bf106-2b0c-4433-9425-754b074e3b4c)

**Comment by glenn-jocher** [comment]
@passerbythesun please keep in mind that the performance of activation functions can vary depending on the dataset and model architecture. While the FReLU activation function may yield promising results in some scenarios, it appears to have limited efficacy in your custom dataset, as indicated by the training procedure stopping at epoch 145 due to no observed improvement in the last 100 epochs. It is recommended to explore other activation functions or further tune the model parameters to achieve better performance on your dataset.

**Comment by IlamSaran** [comment]
Can LeakyReLU be used in YOLOv5x version and how far it will help improve the model (v5x) performance

**Comment by glenn-jocher** [comment]
@IlamSaran LeakyReLU can be used in YOLOv5x and has been shown to improve performance in certain cases, particularly in addressing the vanishing gradient problem. However, the extent of performance improvement may vary depending on the specific dataset and model architecture. It is recommended to experiment with different activation functions and evaluate their impact on model performance to determine the most effective configuration for your use case.

**Comment by IlamSaran** [comment]
HI!
Could you please clarify the difference between Validation set and Test set in training an Deep learning model (Train, Valid???, Test???). And what is testing of unlabeled set of data / unseen set of data that contains similar classes on the custom trained model?

**Comment by glenn-jocher** [comment]
@IlamSaran The validation set is used during training for model selection and hyperparameter tuning, while the test set is reserved for final evaluation of the trained model's performance. Testing on an unlabeled set of data, containing similar classes to those in the training set, helps assess the model's generalization capabilities and its ability to make accurate predictions on previously unseen examples.

**Comment by IlamSaran** [comment]
My DL model for object detection task results with mAP@0.5 = 90% and mAP@0.5:0.95 =78% on my custom created dataset.
But the same model results with 96% for mAP@0.5 and only 55% mAP@0.5:0.95 for a public benchmark dataset. Note (both dataset contain same classes). Though mAP@0.5 is greater for public dataset, mAP@0.5:0.95 is COMPARITBELY VERY LESS COMPARED TO CUSTOM dataset. Please clarify and can i conclude that my model works better on my custom dataset.

**Comment by glenn-jocher** [comment]
@IlamSaran Your model's performance as measured by mAP@0.5 is indeed higher on the public benchmark dataset, indicating good detection at a specific IoU threshold of 0.5. However, the lower mAP@0.5:0.95 score suggests that the model's performance across a range of IoU thresholds from 0.5 to 0.95 is not as robust on the public dataset compared to your custom dataset.

The mAP@0.5:0.95 metric provides a more comprehensive assessment of detection performance across various IoU levels, reflecting both localization and detection accuracy. The comparative decrease in this metric for the public dataset suggests that while the model detects objects well at a lower IoU threshold, it struggles with precise localization at higher IoU thresholds.

In conclusion, your model does appear to work better on your custom dataset when considering the overall detection and localization performance (mAP@0.5:0.95). However, it's also important to consider dataset size, diversity, and difficulty when comparing these metrics.

**Comment by IlamSaran** [comment]
Hi!
I m working on object detection and tracking application using YOLOv5x model. The inference speed for detection is 9.2 ms per frame . Now I integrate SORT tracker with it. How to calculate the FPS to check its real-time video processing  capability.

**Comment by pderrenger** [comment]
To calculate FPS for your integrated detection and tracking application, measure the total processing time per frame (detection + tracking) and use the formula FPS = 1 / total_time_per_frame.
